src/java/org/apache/cassandra/db/virtual/PrimaryIdTable.java

```diff
@@ -0,0 +1,350 @@
+                                     .addRegularColumn(COLUMN_SIZE_ESTIMATE, CounterColumnType.instance)
+                                     .addRegularColumn(COLUMN_SSTABLES, CounterColumnType.instance)
+                                     .build();
+        sizeEstimateColumn = metadata.regularColumns().getSimple(0);
+        sstablesColumn = metadata.regularColumns().getSimple(1);
+    }
+
+    @Override
+    public UnfilteredPartitionIterator select(DecoratedKey partitionKey, ClusteringIndexFilter clusteringIndexFilter, ColumnFilter columnFilter, RowFilter rowFilter)
+    {
+        if (clusteringIndexFilter.isReversed())
+            throw new InvalidRequestException(REVERSED_QUERY_ERROR);
+
+        ByteBuffer[] key = ((CompositeType) this.metadata.partitionKeyType).split(partitionKey.getKey());
+        String keyspace = UTF8Type.instance.getString(key[0]);
+        String table = UTF8Type.instance.getString(key[1]);
+
+        KeyspaceMetadata ksm = Schema.instance.getKeyspaceMetadata(keyspace);
+        if (ksm == null)
+            throw invalidRequest(KEYSPACE_NOT_EXIST_ERROR, keyspace);
+
+        TableMetadata metadata = ksm.getTableOrViewNullable(table);
+        if (metadata == null)
+            throw invalidRequest(TABLE_NOT_EXIST_ERROR, table, keyspace);
+
+        AbstractBounds<PartitionPosition> range = getBounds(metadata, clusteringIndexFilter, rowFilter);
+        return new SingletonUnfilteredPartitionIterator(select(partitionKey, metadata, clusteringIndexFilter, range));
+    }
+
+    private List<SSTableReader> getSStables(TableMetadata metadata, AbstractBounds<PartitionPosition> range)
+    {
+        return Lists.newArrayList(ColumnFamilyStore.getIfExists(metadata).getTracker().getView().liveSSTablesInBounds(range.left, range.right));
+    }
+
+    private UnfilteredRowIterator select(DecoratedKey partitionKey, TableMetadata metadata, ClusteringIndexFilter clusteringIndexFilter, AbstractBounds<PartitionPosition> range)
+    {
+        List<SSTableReader> sstables = getSStables(metadata, range);
+        if (sstables.isEmpty())
+            return UnfilteredRowIterators.noRowsIterator(metadata, partitionKey, Rows.EMPTY_STATIC_ROW, DeletionTime.LIVE, false);
+
+        List<UnfilteredRowIterator> sstableIterators = Lists.newArrayList();
+        for (SSTableReader sstable : sstables)
+            sstableIterators.add(getSStableRowIterator(metadata, partitionKey, sstable, clusteringIndexFilter, range));
+
+        return UnfilteredRowIterators.merge(sstableIterators);
+    }
+
+    private UnfilteredRowIterator getSStableRowIterator(TableMetadata target, DecoratedKey partitionKey, SSTableReader sstable, ClusteringIndexFilter filter, AbstractBounds<PartitionPosition> range)
+    {
+        final KeyReader reader;
+        try
+        {
+            // ignore warning on try-with-resources, the reader will be closed on endOfData or close
+            reader = sstable.keyReader(range.left);
```

Comment:
looking at this usage I wonder if we really need this new method, when we could leverage 

`org.apache.cassandra.io.sstable.format.SSTableReader#keyIterator(org.apache.cassandra.dht.AbstractBounds<org.apache.cassandra.db.PartitionPosition>)`

It's `KeyIterator` rather than `KeyReader`, but should have everything we need.

Reply:
I did originally use key iterator when implementing PoC, thing that was missing was getting the offsets for the start/end between the partitions for calculating size, it just provides the keys. Also worth mentioning in OSS there is no SSTableReader#keyIterator(org.apache.cassandra.dht.AbstractBounds<org.apache.cassandra.db.PartitionPosition>)

Reply:
thanks for the clarification